### ISSUE TITLE
Add answers to warm up data model exercises in rails

### ISSUE THREAD

**Issue by DigitalKnight0** [TheOdinProject/curriculum#23343]
In the micro-reddit and private events project, The warmup part should have some answers linked so the users can compare their solutions to check whether they got it right or not. This helps in understanding associations better as they can research and rectify their mistake before starting the project. Currently, the only option they have is to post their solutions in the discord and ask for feedback, i think it might be better to link a repo or something with the correct answers

**Comment by rlmoser99** [comment]
I do not think providing a repo with the answers is the best solution, because there can be different ways to implement associations. One thought that I've had in the past, but haven't developed is to create a test-driven approach to this. If the tests pass, you can be certain that you've written the associations correctly. 

Is this something you'd be interested into working on?

**Comment by DigitalKnight0** [comment]
@rlmoser99 I would love to, but i am myself currently on private events project and quite unsure of how testing is done with rails. I can prolly get started with some guidance but still your call

**Comment by rlmoser99** [comment]
Ok. Before we decide on anything, let me touch base with the rest of the Rails team and see where this fits in with our current plans. Thanks for creating an issue to help find a way to provide feedback on those association warm ups. 👍

**Comment by dm-murphy** [comment]
If you're seeking contributors to help add a test-driven approach here or elsewhere in the Rails curriculum, I'd love to help! :-)

**Comment by rlmoser99** [comment]
@DigitalKnight0 @dm-murphy Quick Update: I've talked with a few other Rails maintainers and we think the TDD approach would be the best direction to go. The tests that I have mind (and have some sample ones written) will **not** be the typical [Rails association tests](https://www.codewithjason.com/examples-pointless-rspec-tests/), because they give away the answers... and are in general pretty pointless! 

However, we don't want to create 8 new repos for Rails applications just for these warm-up exercises. So, it was suggested to use [Rails Templates](https://guides.rubyonrails.org/rails_application_templates.html) for this project. I am in the middle of another project, so it will take me some time to investigate how best to move forward with using templates and then I'll create issues to turn each warm-up exercise into a template that provides test files.

In the meantime, as you go through the Rails curriculum, I suggest learning about testing with [codewithjason](https://www.codewithjason.com/articles/) and [FactoryBot](https://github.com/thoughtbot/factory_bot).

**Comment by dm-murphy** [comment]
> @DigitalKnight0 @dm-murphy Quick Update: I've talked with a few other Rails maintainers and we think the TDD approach would be the best direction to go. The tests that I have mind (and have some sample ones written) will **not** be the typical [Rails association tests](https://www.codewithjason.com/examples-pointless-rspec-tests/), because they give away the answers... and are in general pretty pointless!
> 
> However, we don't want to create 8 new repos for Rails applications just for these warm-up exercises. So, it was suggested to use [Rails Templates](https://guides.rubyonrails.org/rails_application_templates.html) for this project. I am in the middle of another project, so it will take me some time to investigate how best to move forward with using templates and then I'll create issues to turn each warm-up exercise into a template that provides test files.
> 
> In the meantime, as you go through the Rails curriculum, I suggest learning about testing with [codewithjason](https://www.codewithjason.com/articles/) and [FactoryBot](https://github.com/thoughtbot/factory_bot).

Thanks for the update!

I'm almost done working through The Complete Guide to Rails Testing from codewithjason and I've found it really helpful and clear. Both the free content from the site and the full book could be great resources in the curriculum.

**Comment by CouchofTomato** [comment]
> I'm almost done working through The Complete Guide to Rails Testing from codewithjason and I've found it really helpful and clear. Both the free content from the site and the full book could be great resources in the curriculum.

I agree although the full book is a paid resource which means we could only at most add it to the additional resources section

**Comment by rlmoser99** [comment]
@TheOdinProject/rails-path I am not sure we have this idea documented with our newer plans for rails updates. I still very much want to see this, but it just hasn't been a top priority.

https://github.com/TheOdinProject/top-meta/issues/173

**Comment by JoshDevHub** [comment]
Resurrecting this for a bit.

I have several thoughts:
1. I think it'd be a good idea to construct an answer key for this. It's a common question in the discord with learners reaching this and being uncertain they've come to the correct conclusions.
1. The TDD idea is interesting, but I think it has a couple of drawbacks that might be difficult to get around. For one, even if you don't write association tests using shoulda matchers, I think it's likely challenging to hide what the answers would be from learners who read the test source code. Maybe there are ways to do it that I'm not thinking of? You might be able to hide certain things around associations, but I don't think there's going to be a good way to hide what models exist--and that's a big part of the battle on stuff like this to me. I also think there's no way to avoid it prescribing a very specific solution (use *these* models with *these* names and *these* attributes and *these* associations, or your tests fail)--but data modeling is looser than that. There are often multiple valid ways to construct the data, and there are definitely multiple valid ways to name entities and their relationships.
1. Data modeling can (and in my opinion, often *should*) be worked out on paper (well "paper"--or a whiteboard or excalidraw or an ERD or whatever) without Rails or writing any code or anything. I don't see these data model exercises as something learners need a Rails app for. I'm sure some people reason through these things differently, but for me, it's always been beneficial to iron this stuff out away from a strictly Rails/ActiveRecord context.
1. However, even if people are really set on getting to a TDD solution eventually, I think it could be good to have a stopgap type thing like Chook mentioned above.

**Comment by grimley517** [comment]
Not sure what an answer-key is in the context of above comment.

What may be more natural might be the inclusion of some detailed acceptance criteria. Acceptance criteria are used  extensively in real projects for engineers to evaluate whether their tickets are done enough to be considered complete.
